README.md
# Enpass-JSON-to-Bitwarden-converter-fix
Python script to fix the custom fields not being show as hidden on Bitwarden when importing Enpass JSON


## Summary
When importing an Enpass JSON file to Bitwarden, if you have custom fields that are hidden/marked as "Sensitive" on Enpass, they are imported as standard text fields in Bitwarden, not as hidden (tested on Version 2.13.2)

![Enpass vs Bitwarden](images/Enpass_hidden_vs_Bitwarden.jpg) 


The python script will compare an Enpass JSON file with a Bitwarden JSON file. If it finds fields that should be hidden, it will correct the Bitwarden JSON. You can then import this new file to your vault.

DISCLAIMER: ALWAYS keep a backup of the JSON files in a secure place. This files will have all your passwords in plain text, so keep them in safe place and consider encrypting them/and put them in password protected zip. 
I do not take resposability of this script messing up or loosing info of your vault. 

## Pre-requisites
- Python3 (probably will work with Python2.7 or will need small adjustments)
- JSON python library (not sure if included by default on Python3 distribution)

This is how Enpass entry looks like, with sensitive information hidden:

![Enpass hidden example](images/Enpass_hidden_example.jpg?raw=true)

## Instructions
1. Export your Enpass vault as JSON. Go to File -> Export. Select JSON and save the file to your desktop. Copy the file to the working directly and rename it to Enpass_vault.json
2. Open Bitwarden web https://vault.bitwarden.com/ (or your own self-hosted bitwaren) and login to with your username.
3. Navigate to Tools → Import Data. Choose from the dropdown Enpass(JSON) and select your previous created file. Upload it.
4. Check some of the items. On the photo below you can observe how the custom "Sensitive/password" fields at the bottom are not hidden with *** because of the incorrect import.

![Bitwarden wrong import](images/Bitwarden_imported_incorrectly.jpg) 

5. Go again to Tools → Export Vault.Choose JSON and put your Bitwarden password.
6. Copy the saved file (should have a format similar to "bitwarden_export_20200426121443.json" to the working folder and rename it as Bitwarden_exported.json
  IMPORTANT: COPY THE FILE, do not MOVE IT. Always keep the original backup Bitwarden file. 
7. Execute the python script. It will create a new Bitwarden_fixed.json file
8. You can check the output file Fixed_DELETE.txt to see which fields have been fixed. You can also compare the Bitwarden_exported.json with the Bitwarden_fixed.json file.

  **IMPORTANT: DELETE the Fixed_DELETE.txt file since it will have plain text visible password!**

9. Delete your Bitwarden vault. Go to Settings → At the bottom of the page, in the "Danger zone" box, click "Purge vault"
10. Import now the corrected file name Bitwarden_fixed.json Go to Tools → Import Data. Choose from the dropdown the first element, BITWARDEN(JSON) and select the Bitwarden_fixed.json file. Upload it.
11. Check again the items. The custom fields should now be correctly hidden when appropiate.

![Bitwarden fixed](https://github.com/jvteleco/Enpass-JSON-to-Bitwarden-converter-fix/blob/master/images/Bitwarden_fields_fixed.jpg?raw=true) 


**REMEMBER to delete the Fixed_DELETE.txt file once you have finished and move all the JSON files to a secure place since they will have all your passwords.**


Changelog.md
# Changelog
All notable changes to this project will be documented in this file.


## [0.0.1] - 2020-04-26
### Added
- Created Changelog.md
- First version of python script

### Changed

### Removed


EnpassJson_Bitwarden_converter.py
# -*- coding: utf-8 -*-

#-----------------------------------------------------------------------
# EnpassJson_Bitwarden_converter.py
# AUTHOR: JVTELECO
# v1.0 26/04/2020
#-----------------------------------------------------------------------

"""
This EnpassJson_Bitwarden_converter.py script allows to fix the hidden fields of 
the Bitwarden Json file after importing an Enpass JSON.
"""



import traceback
import json


EnpassFileName="Enpass_vault.json"
BitwardenFileName="Bitwarden_exported.json"

items_processed = 0
items_processed_changed = 0

print("Opening Enpass file")
with open(EnpassFileName, encoding="utf8") as f:
    json_data_enpass = json.load(f)
  
##print(json_data)  
#print(json.dumps(json_data_enpass, indent = 2))
##print(json.dumps(json_data, indent = 2, sort_keys=True))

print("Opening Bitwarden file")
with open(BitwardenFileName,encoding="utf8") as f_bt:
    json_data_bitwarden = json.load(f_bt)
  

  
fixed_file= open('Fixed_DELETE.txt', 'w')
  
##print(json_data)  
#print(json.dumps(json_data_bitwarden, indent = 2))
##print(json.dumps(json_data, indent = 2, sort_keys=True))




long_json_data_enpass_items = len(json_data_enpass["items"])
long_json_data_bitwarden_items = len(json_data_bitwarden["items"])
print("Enpass number items:", long_json_data_enpass_items )
print("Bitwarden number items:", long_json_data_bitwarden_items)

print("\n\n Starting comparison...\n\n")

for i in range(0, long_json_data_enpass_items):

    item_enpass = json_data_enpass["items"][i]
    item_enpass_title = item_enpass["title"]
#    print("\nENPASS:\t\t", item_enpass_title)
    
    for j in range(0, long_json_data_bitwarden_items):
        item_bitwarden = json_data_bitwarden["items"][j]
        item_bitwarden_name = item_bitwarden["name"]
        if item_enpass_title == item_bitwarden_name:
#            print("Bitwarden:\t", item_bitwarden_name)
            #Now check for sensitive items
            #print(json.dumps(item_enpass, indent = 2))
            #print(json.dumps(item_bitwarden, indent = 2))
            
            #we now only the items in the fields that are the custom fields
            try:
#            if (1):
                for k in range(0, len(item_bitwarden["fields"])):
                    field_bitwarden=item_bitwarden["fields"][k]
                    #print(field_bitwarden)
                    
                    
                    #we now check the name against the Enpass item
                    #when there is a match, need to check the sensitive value
                    #if it is '1', need to change the bitwarden item type to '1' so it is hidden
                    for m in range(0, len(item_enpass["fields"])):
                        field_enpass=item_enpass["fields"][m]
                        if field_bitwarden["name"] == field_enpass["label"]:
#                            print("MATCH")
#                            print(field_bitwarden)
#                            print(field_enpass)
                            #print(type(field_enpass["sensitive"]))
                            #print(type(field_bitwarden["type"]))
                            if field_enpass["sensitive"] == 1:
                                field_bitwarden["type"] = 1
                                items_processed_changed = items_processed_changed + 1
                                #print("CHANGED:\t", field_bitwarden)
                                fixed_file.write(str(item_bitwarden_name))
                                fixed_file.write("\t")
                                fixed_file.write(str(field_bitwarden))
                                fixed_file.write("\n")
                
                
                
                
                
                items_processed=items_processed+1
                #end clause of: if item_enpass_title == item_bitwarden_name:
            except Exception as e:
                #probably if a bitwarden does not have custom fields, will throw an exception of "fields" KeyError, does not exist
                ##print(e)
#                print("ERROR Bitwarden:\t", item_bitwarden_name)
#                print(traceback.format_exc())
                pass
                
                
print("\n\n Finised comparison.\n\n")



print("Enpass number items:", long_json_data_enpass_items )
print("Bitwarden number items:", long_json_data_bitwarden_items)
print("Number of items processed with custom fields", items_processed)
print("Number of fields fixed to hidden", items_processed_changed)




with open('Bitwarden_fixed.json', 'w') as json_file:
    json.dump(json_data_bitwarden, json_file, indent = 2)
  
  
fixed_file.close()
